File: kotlin-node/karakum/src/jsMain/resources/annotations/annotateDuplex.js
import ts from "typescript"

export default (node) => {
    const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"

    if (
        sourceFileName.endsWith("stream.d.ts")
        && ts.isClassDeclaration(node)
        && node.name?.text === "Duplex"
    ) {
        return `@Suppress("MANY_CLASSES_IN_SUPERTYPE_LIST")`
    }

    if (
        sourceFileName.endsWith("stream.d.ts")
        && ts.isExpressionWithTypeArguments(node)
        && ts.isIdentifier(node.expression)
        && (
            node.expression.text === "Readable"
            || node.expression.text === "Writable"
        )

        && node.parent
        && ts.isHeritageClause(node.parent)

        && node.parent.parent
        && ts.isInterfaceDeclaration(node.parent.parent)
        && node.parent.parent.name.text === "Duplex"
    ) {
        return `@seskar.js.JsMixin`
    }

    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/wrappersgenerator/node/annotations/annotateDuplex.kt
package wrappersgenerator.node.annotations

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.AnnotationContext
import io.github.sgrishchenko.karakum.util.getParentOrNull
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import typescript.*

fun annotateDuplex(node: Node, context: AnnotationContext) = nullable {
    val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName

    ensure(sourceFileName.endsWith("stream.d.ts"))

    nullable {
        ensure(isClassDeclaration(node))
        ensure(node.name?.text == "Duplex")

        "@Suppress(\"MANY_CLASSES_IN_SUPERTYPE_LIST\")"
    } ?: nullable {
        ensure(isExpressionWithTypeArguments(node))

        val expression = node.expression
        ensure(isIdentifier(expression))
        ensure(
            expression.text == "Readable"
                    || expression.text == "Writable"
        )

        val heritageClause = ensureNotNull(node.getParentOrNull())
        ensure(isHeritageClause(heritageClause))

        val interfaceNode = ensureNotNull(heritageClause.getParentOrNull())
        ensure(isInterfaceDeclaration(interfaceNode))
        ensure(interfaceNode.name.text == "Duplex")

        "@seskar.js.JsMixin"
    }
}
